### PR TITLE
Port actual OpenGL implementation for RenderEngineGl

### PR DESCRIPTION
**Pull request by tehbelinda**
- First commit is code copied back in from Anzu (except for a few merges in gl_renderer BUILD.bazel)
- Second commit has remaining changes for Drake compatibility

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/12788)

### CODE REVIEW

**Comment by tehbelinda** [comment]
take two: +@seancurtis-tri for feature review please

**Comment by jwnimmer-tri** [comment]
I only did a quick skim, but I didn't see a plan for handling the oversize PR?  Make sure you have a plan for spitting before you get too far along?

**Comment by tehbelinda** [comment]
___
*[bindings/pydrake/BUILD.bazel, line 423 at r4](https://reviewable.io/reviews/robotlocomotion/drake/12788#-M1RJWE792alN6BBsIwR:-M1hHZht6TJ36QSbtCIL:b-15hs9p) ([raw file](https://github.com/robotlocomotion/drake/blob/c166a36b70bfbc2d9df12bac05f2d9eecc76f0db/bindings/pydrake/BUILD.bazel#L423)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Why has this become manual?
</blockquote></details>

I added a test for the bound GL render engine where there was already a test for an existing render engine, so now that it pulls it in, it needs the same manual tag treatment... do you think it's better to split the render engines into a separate test, geometry_render_engine_test maybe?

**Comment by tehbelinda** [comment]
___
*[geometry/render/gl_renderer/BUILD.bazel, line 123 at r4](https://reviewable.io/reviews/robotlocomotion/drake/12788#-M1QuqBnDEhvrXP8TdaK:-M1hIjQBDwmDDq6MP096:b-pw77dj) ([raw file](https://github.com/robotlocomotion/drake/blob/c166a36b70bfbc2d9df12bac05f2d9eecc76f0db/geometry/render/gl_renderer/BUILD.bazel#L123)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Why strictly manual?

Also, what's the effect of `local = 1` compared to the "manual" tag?
</blockquote></details>

I just used it as suggested from https://github.com/RobotLocomotion/drake/pull/7004, let me try it without it again to be sure... if it does help I'll add the issue link for future reference

**Comment by tehbelinda** [comment]
___
*[geometry/render/gl_renderer/load_mesh.h, line 15 at r4](https://reviewable.io/reviews/robotlocomotion/drake/12788#-M1QviVN9mErOZcSH9Bm:-M1jlVg501sxK2PLaFRQ:buqapsh) ([raw file](https://github.com/robotlocomotion/drake/blob/c166a36b70bfbc2d9df12bac05f2d9eecc76f0db/geometry/render/gl_renderer/load_mesh.h#L15)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Can you put a huge TODO on this suggesting that we remove this functionality in favor of the "load OBJ as `SurfaceMesh`" functionality? It'll require some refactoring down the road, but it's worth noting that we have redundant functionality now.
</blockquote></details>

Sure. But does that mean I should be adding the reading in of normals into there if that's the end goal?

**Comment by tehbelinda** [comment]
___
*[geometry/render/gl_renderer/render_engine_gl.cc, line 225 at r3](https://reviewable.io/reviews/robotlocomotion/drake/12788#-M1R5pU4C96s2tggdwqr:-M1mWMHb2j4mtbYws49N:b-5c3ywd) ([raw file](https://github.com/robotlocomotion/drake/blob/855cb11283e5adca87de456ec07a67d2900c58bf/geometry/render/gl_renderer/render_engine_gl.cc#L225)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

nit: Is this a meaningful action? In `SetCameraProperties` we attach the frame buffer/texture to `GL_COLOR_ATTACHMENT0`. In the shader, we are writing to `layout(location = 0)`.
</blockquote></details>

Yes. I believe that what you are referring to is separate. Apparently depth calculation happens automatically on the GPU, separately to whatever the fragment shader wants to do, so that depth buffer attachment is still needed to do the calculation. I'm not sure how clear that sounds so here is a link to a StackOverflow thread that further exemplifies the separation between fragment/color shader and depth: https://stackoverflow.com/a/23372000

I did some more refresher reading (https://learnopengl.com/Advanced-OpenGL/Depth-testing) and it's still needed for a correct render since `OpenGL performs a depth test and if this test passes, the fragment is rendered and the depth buffer is updated with the new depth value. If the depth test fails, the fragment is discarded.`

Also I commented it out to see what would happen and the tests fail because the depths are off

**Comment by tehbelinda** [comment]
___
*[bindings/pydrake/BUILD.bazel, line 423 at r4](https://reviewable.io/reviews/robotlocomotion/drake/12788#-M1RJWE792alN6BBsIwR:-M23bXFrBalaCieRZZYY:b-w9d5fg) ([raw file](https://github.com/robotlocomotion/drake/blob/c166a36b70bfbc2d9df12bac05f2d9eecc76f0db/bindings/pydrake/BUILD.bazel#L423)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

To my shame, I didn't follow that. :-/
</blockquote></details>

Let me try again...
So I think it's because I added a test for the Python binding of RenderEngineGl to the rendering section of geometry_test.py
```
        # Test limited rendering API.
        scene_graph.AddRenderer("test_renderer",
                                mut.render.MakeRenderEngineVtk(
                                    mut.render.RenderEngineVtkParams()))
        self.assertTrue(scene_graph.HasRenderer("test_renderer"))
        self.assertEqual(scene_graph.RendererCount(), 1)
        if 'darwin' in sys.platform:  # OpenGL is not supported on macOS.
            self.assertRaises(RuntimeError, mut.render.MakeRenderEngineGl)
        else:
            scene_graph.AddRenderer("gl_renderer",
                                    mut.render.MakeRenderEngineGl())
            self.assertTrue(scene_graph.HasRenderer("gl_renderer"))
            self.assertEqual(scene_graph.RendererCount(), 2)
```

Render_engine_gl_test pulls in RenderEngineGl and its dependencies and thus has to be run via manual. So I believe that similarly, the geometry_test.py now also pulls in RenderEngineGl and thus needs the manual tag

I was asking if it made sense to move the rendering part of the test out, e.g. to a geometry_render_engine_test.py, so that only that test is marked manual, and not all of geometry_test.py?

**Comment by tehbelinda** [comment]
___
*[bindings/pydrake/BUILD.bazel, line 423 at r4](https://reviewable.io/reviews/robotlocomotion/drake/12788#-M1RJWE792alN6BBsIwR:-M241zlZ7vvTGYk7Kxc5:b-synvli) ([raw file](https://github.com/robotlocomotion/drake/blob/c166a36b70bfbc2d9df12bac05f2d9eecc76f0db/bindings/pydrake/BUILD.bazel#L423)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

So, let me see if I understand this:

1. This is `manual` because if it isn't, it gets run in CI.
2. If it gets run in CI, it gets run for a mac.
3. If it gets run for a mac, we throw an exception.

So, what was the reasoning for throwing for `darwin` rather than just skipping the test for `RenderEngineGL`?
</blockquote></details>

Yup
To test that it correctly throws on mac even in the python version? I can remove it if you think it's not necessary

**Comment by tehbelinda** [comment]
___
*[bindings/pydrake/BUILD.bazel, line 423 at r4](https://reviewable.io/reviews/robotlocomotion/drake/12788#-M1RJWE792alN6BBsIwR:-M24JwseBt2ibgvNx6G_:b-5yprcf) ([raw file](https://github.com/robotlocomotion/drake/blob/c166a36b70bfbc2d9df12bac05f2d9eecc76f0db/bindings/pydrake/BUILD.bazel#L423)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

And yet, throwing an exception is fine in that test. It can run on mac CI and we *expect* it to throw an exception and the framework handles it fine. So, manual isn't necessary; running in CI is fine, yes?
</blockquote></details>

But for the Ubuntu branch, manual is necessary?

**Comment by tehbelinda** [comment]
___
*[bindings/pydrake/BUILD.bazel, line 423 at r4](https://reviewable.io/reviews/robotlocomotion/drake/12788#-M1RJWE792alN6BBsIwR:-M24Kmmb70jOUozkr-G9:b-2ptg1x) ([raw file](https://github.com/robotlocomotion/drake/blob/c166a36b70bfbc2d9df12bac05f2d9eecc76f0db/bindings/pydrake/BUILD.bazel#L423)):*
<details><summary><i>Previously, tehbelinda (Bel) wrote…</i></summary><blockquote>

But for the Ubuntu branch, manual is necessary?
</blockquote></details>

Running in CI is fine for Mac but not fine for Ubuntu

**Comment by tehbelinda** [comment]
___
*[bindings/pydrake/BUILD.bazel, line 423 at r4](https://reviewable.io/reviews/robotlocomotion/drake/12788#-M1RJWE792alN6BBsIwR:-M24nv5-1SnNMsLjNym8:bt47pp6) ([raw file](https://github.com/robotlocomotion/drake/blob/c166a36b70bfbc2d9df12bac05f2d9eecc76f0db/bindings/pydrake/BUILD.bazel#L423)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Why can we run the vtk and ospray tests in CI? But not this one?
</blockquote></details>

I was about to say - they have the manual tag too, but I see what you're saying now...the existing python one had Vtk before the manual tag. Ok let me try it

**Comment by tehbelinda** [comment]
___
*[bindings/pydrake/BUILD.bazel, line 423 at r4](https://reviewable.io/reviews/robotlocomotion/drake/12788#-M1RJWE792alN6BBsIwR:-M24vbKd-RM0GSK6IocD:b-7wha6s) ([raw file](https://github.com/robotlocomotion/drake/blob/c166a36b70bfbc2d9df12bac05f2d9eecc76f0db/bindings/pydrake/BUILD.bazel#L423)):*
<details><summary><i>Previously, tehbelinda (Bel) wrote…</i></summary><blockquote>

I was about to say - they have the manual tag too, but I see what you're saying now...the existing python one had Vtk before the manual tag. Ok let me try it
</blockquote></details>

Ok I think it must be because unlike Vtk/Ospray, the Gl one depends on X, which is what the tests are complaining about
